Add div and mod, function calling support to WC_Eval_Math (#27730)

cpiber opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Div and mod capability to math class. Complete function capability (calling not supported). Add tests.

Closes #27730

### How to test the changes in this Pull Request:

In shipment costs, add equation with `%` or `\`

Go to WooCommerce > Settings > Shipping > Locations not covered by your other zones > Add shipping method: Flat rate. Edit and set cost to `10 + 5 * ([qty] \\ 10)`. Put a product into cart and change quantities: The formula sets a base price of 10 € and adds 5 € per completed 10 products.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?

<!-- Mark completed items with an [x] -->

cpiber committed:
Add div and mod support to WC_Eval_Math (#27730)

cpiber commented:
I can also add tests for this class if you want me to, I did not find any

rrennick commented:
> I can also add tests for this class if you want me to

@cpiber We would appreciate it if you did. Also, it would help with the review process if you edited the PR description and provided a sample and/or instructions for testing.

cpiber commented:
@rrennick I updated the description, working on the tests now.
While doing that I noticed that `stripslashes` is used on the string, so `\` simply gets removed (the symbol for integer division). What would you recommend to do in this case? I see two options: changing the code that saves the formula (is unslash even needed?) or choose a different symbol.

rrennick commented:
> While doing that I noticed that stripslashes is used on the string, so \ simply gets removed (the symbol for integer division). What would you recommend to do in this case?

@cpiber Backslashes need to be escaped in the input string. Characters are escaped with a backslash.  Because you are escaping the escape character you would use `\\\\`.

cpiber commented:
@rrennick `\\` is correct, I only need one backslash. The problem is that simply using `\` does not work (it is stripped), and even using `\\` is stripped to `\` when saving, thus when editing one has to manually add a second `\` again.

cpiber committed:
WC_Eval_Math add missing function evaluation

Previously supported setting but not evaluating functions
Added $suppress to override not triggering errors

cpiber committed:
Tests for WC_Eval_Math

cpiber commented:
@rrennick I added tests, please let me know if there are more tests you'd like to add or if you think they could be done better.

I also added function calling support because I noticed that while defining a function is possible, there was no way to call it. Single-argument functions would just return their argument, multi-argument functions would generate an invalid stack (missing operation).

cpiber commented on includes/libraries/class-wc-eval-math.php:
```diff
@@ -282,24 +290,57 @@ private static function pfx( $tokens, $vars = array() ) {
 				} elseif ( ! preg_match( "/^([a-z]\w*)\($/", $token, $matches ) ) {
 					if ( is_numeric( $token ) ) {
-						$stack->push( $token );
+						$stack->push( $token + 0 );
```
I added this to make sure that only numbers are on the stack. While this is normally not needed (operations convert automatically), e.g. `evaluate("2")` unexpectedly returned a string.

rrennick replied:
`$token + 0` will probably give a warning in PHP 7.4 when `$token` is a string. It would be better to cast `$token` to an `int` or `float`.

cpiber replied:
`$token` was always a string, the only thing that changed is that I explicitly do it here instead of above where the operators are evaluated, so the warning was there before if there was one.
So unconditionally convert to float? Or should I try to convert to int if possible else float?

cpiber committed:
Fix typo

cpiber commented:
@rrennick I'd like to finish this before the weekend, can you please review?

rrennick commented:
> The problem is that simply using \ does not work (it is stripped), and even using \\ is stripped to \ when saving, thus when editing one has to manually add a second \ again.

This could be fixed in the field saving code.

rrennick commented:
> I added tests, please let me know if there are more tests you'd like to add or if you think they could be done better.

Someone on the team will follow up on this when reviewing the PR.

cpiber commented:
> > The problem is that simply using \ does not work (it is stripped), and even using \ is stripped to \ when saving, thus when editing one has to manually add a second \ again.
> 
> This could be fixed in the field saving code.

Yes, as I was saying. Problem is that the field is unslashed and that the `\` operator cannot be reliably distinguished from one that is supposed to escape. Though I have no idea what one could want to escape. That's why I proposed to find a different symbol. Unless you have an idea how to implement that.

cpiber commented:
Any chance of anything happening?

cpiber commented:
@rrennick I guess this is dead?

rrennick commented:
@cpiber I just flagged this for someone on the team to review. Could you merge current trunk into it to trigger running the automated tests?

cpiber committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce
